pkg/libdevfile/libdevfile.go

```diff
@@ -119,3 +120,68 @@ func execDevfileEvent(devfileObj parser.DevfileObj, events []string, handler Han
 	}
 	return nil
 }
+
+// GetContainerComponents returns a slice of container components in the given devfile
+func GetContainerComponents(devfileObj parser.DevfileObj) ([]v1alpha2.Component, error) {
+	return devfileObj.Data.GetComponents(common.DevfileOptions{
+		ComponentOptions: common.ComponentOptions{ComponentType: v1alpha2.ContainerComponentType},
+	})
+}
+
+// GetContainerEndpointMapping returns a map of container names and slice of its endpoints (in int) with exposure status other than none
+func GetContainerEndpointMapping(containers []v1alpha2.Component) map[string][]int {
+	ceMapping := make(map[string][]int)
+	for _, container := range containers {
+		if container.ComponentUnion.Container == nil {
+			// this is not a container component; continue prevents panic when accessing Endpoints field
+			continue
+		}
+		k := container.Name
+		if _, ok := ceMapping[k]; !ok {
+			ceMapping[k] = []int{}
+		}
+
+		endpoints := container.Container.Endpoints
+		for _, e := range endpoints {
+			if e.Exposure != v1alpha2.NoneEndpointExposure {
+				ceMapping[k] = append(ceMapping[k], e.TargetPort)
+			}
+		}
+	}
+	return ceMapping
+}
+
+// GetPublicAndInternalEndpointsFromDevfile returns a slice of all endpoints in a devfile with exposure value not set to public or internal
+func GetPublicAndInternalEndpointsFromDevfile(devfileObj parser.DevfileObj) ([]v1alpha2.Endpoint, error) {
+	containers, err := GetContainerComponents(devfileObj)
+	if err != nil {
+		return nil, err
+	}
+
+	var endpoints []v1alpha2.Endpoint
+	for _, c := range containers {
+		for _, e := range c.Container.Endpoints {
+			if e.Exposure == v1alpha2.PublicEndpointExposure || e.Exposure == v1alpha2.InternalEndpointExposure {
+				endpoints = append(endpoints, e)
+			}
+		}
+
+	}
+	return endpoints, nil
+}
+
+// HaveEndpointsChanged returns true if the total number of public and/or internal endpoints have changed between
+// the devfile objects
+func HaveEndpointsChanged(oldDevfile, newDevfile parser.DevfileObj) (bool, error) {
```

Comment:
If you want to make this function more generic for future use, you could have a `HaveChanged` function instead, that would return a set of booleans (either through a returned struct of a list of return values). Choosing a struct as return value would help to not change caller's codes when you add more fields into the struct

Reply:
Makes sense, but can we do that in a separate PR when/if we really figure a need to make this function more generic?